contrib/mesos/pkg/minion/server.go

```diff
@@ -0,0 +1,223 @@
+	proxyLogfile string
+	proxyBindall bool
+}
+
+func NewMinionServer() *MinionServer {
+	s := &MinionServer{
+		KubeletExecutorServer: exservice.NewKubeletExecutorServer(),
+		privateMountNS:        true,
+		done:                  make(chan struct{}),
+
+		runProxy:     true,
+		proxyLogfile: "./proxy.log",
+	}
+
+	// cache this for later use
+	binary, err := osext.Executable()
+	if err != nil {
+		log.Fatalf("failed to determine currently running executable: %v", err)
+	}
+	s.kmBinary = binary
+
+	return s
+}
+
+// filterArgsByFlagSet returns a list of args which are parsed by the given flag set
+// and another list with those which do not match
+func filterArgsByFlagSet(args []string, flags *pflag.FlagSet) ([]string, []string) {
+	matched := []string{}
+	notMatched := []string{}
+	for _, arg := range args {
+		err := flags.Parse([]string{arg})
+		if err != nil {
+			notMatched = append(notMatched, arg)
+		} else {
+			matched = append(matched, arg)
+		}
+	}
+	return matched, notMatched
+}
+
+func (ms *MinionServer) launchProxyServer() {
+	bindAddress := "0.0.0.0"
+	if !ms.proxyBindall {
+		bindAddress = ms.KubeletExecutorServer.Address.String()
+	}
+	args := []string{
+		fmt.Sprintf("--bind-address=%s", bindAddress),
+		fmt.Sprintf("--v=%d", ms.proxyLogV),
+		"--logtostderr=true",
+	}
+
+	if ms.clientConfig.Host != "" {
+		args = append(args, fmt.Sprintf("--master=%s", ms.clientConfig.Host))
+	}
+
+	ms.launchHyperkubeServer(hyperkube.KM_PROXY, &args, ms.proxyLogfile)
+}
+
+func (ms *MinionServer) launchExecutorServer() {
+	allArgs := os.Args[1:]
+
+	// filter out minion flags, leaving those for the executor
+	executorFlags := pflag.NewFlagSet("executor flags", pflag.ContinueOnError)
+	executorFlags.SetOutput(ioutil.Discard)
+	ms.AddExecutorFlags(executorFlags)
+	executorArgs, _ := filterArgsByFlagSet(allArgs, executorFlags)
+
+	// run executor and quit minion server when this exits cleanly
+	err := ms.launchHyperkubeServer(hyperkube.KM_EXECUTOR, &executorArgs, "executor.log")
+	if err == nil {
+		log.Info("Executor terminated cleanly, stopping the minion")
+		close(ms.done)
+	}
+}
+
+func (ms *MinionServer) launchHyperkubeServer(server string, args *[]string, logFileName string) error {
+	log.Infof("Spawning hyperkube %v with args '%+v'", server, args)
+
+	// prepare parameters
+	kmArgs := []string{server}
+	for _, arg := range *args {
+		kmArgs = append(kmArgs, arg)
+	}
+
+	// create command
+	cmd := exec.Command(ms.kmBinary, kmArgs...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		log.Fatal(err)
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ch := make(chan struct{})
+	go func() {
+		defer func() {
+			select {
+			case <-ch:
+				log.Infof("killing %v process...", server)
+				if err = cmd.Process.Kill(); err != nil {
+					log.Errorf("failed to kill %v process: %v", server, err)
+				}
+			default:
+			}
+		}()
+
+		writer := lumberjack.Logger{
+			Filename:   logFileName,
+			MaxSize:    1,
```

Comment:
magic numbers (1, 3). extract constants (as defaults) and allow configuration via params.